### PR TITLE
8316304: (fs) Add support for BasicFileAttributes.creationTime() for Linux

### PR DESCRIPTION
**Pull request by jerboaa**
This is an unclean backport of the filesystem enhancement on Linux to use `statx` where available. The JDK 21u patch didn't apply cleanly because JDK 20+ has [JDK-8283335](https://bugs.openjdk.org/browse/JDK-8283335) which isn't needed for this backport. The relevant code has been adjusted to work without that enhancement being present (see different commits). Copyright dates needed manual adjusting and the test needed some trivial changes related to test lib and import so that it works (failed to compile on 17u).

As for the code changes in `UnixNativeDispatcher.c`, the JNI method `Java_sun_nio_fs_UnixNativeDispatcher_stat1()` returning the mode for a file/directory doesn't exist in JDK 21 (again [JDK-8283335](https://bugs.openjdk.org/browse/JDK-8283335)). Therefore, I've adjusted the code to also use `statx` where available for symmetry.

I'll follow up with the related build fix (for the test libs) in a dependent PR (JDK-8319922).

Please review!

Testing:
- [x] GHA
- [x] Build test on various Linux (aarch64, Alpine x64, ppc64le, x64) and Unix OSes (AIX, Mac) so as to not break the build there.
- [x] Manual testing for `BasicFileAttributes.creationTime()`. Looks good.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] [JDK-8316304](https://bugs.openjdk.org/browse/JDK-8316304) needs maintainer approval
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8316304](https://bugs.openjdk.org/browse/JDK-8316304): (fs) Add support for BasicFileAttributes.creationTime() for Linux (**Enhancement** - P3 - Approved)

### Reviewers
 * [Thomas Stuefe](https://openjdk.org/census#stuefe) (@tstuefe - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk17u-dev.git pull/2175/head:pull/2175` \
`$ git checkout pull/2175`

Update a local copy of the PR: \
`$ git checkout pull/2175` \
`$ git pull https://git.openjdk.org/jdk17u-dev.git pull/2175/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 2175`

View PR using the GUI difftool: \
`$ git pr show -t 2175`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk17u-dev/pull/2175.diff">https://git.openjdk.org/jdk17u-dev/pull/2175.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk17u-dev/pull/2175#issuecomment-1910407514)

### CODE REVIEW

**Review by tstuefe — COMMENTED** [review]
One thing I wonder is whether there could be cases where statx may fail but stat64 may succeed; in that case, we would have a regression now. Looking at the manpages, it seems that the set of errno values is the same for both, so it may be a nonissue. If we wanted to be super careful though, we could always retry stat64 if statx fails. Would also shorten the functions a bit.

Up to you, especially since that would introduce changes in the downport release.

I'm reviewing now...

**Review by tstuefe — APPROVED** [review]
Looks good. I wonder about backward compatibility - whether there is code that relies on creation time being the modification time. If so, its an error of course, but still may cause problems. This will be mentioned in a release note, or?

**Comment by jerboaa** [comment]
Thanks for the review!

> I wonder about backward compatibility - whether there is code that relies on creation time being the modification time. If so, its an error of course, but still may cause problems.

Users already get different behaviour on say, Mac OS, Windows and Linux (with statx support) today. The former two support `creationTime()`, Linux does not (prior this patch). The [javadoc](https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/nio/file/attribute/BasicFileAttributes.html#creationTime()) from creationTime() mentions:

```
If the file system implementation does not support a time stamp to indicate the time when the file was
created then this method returns an implementation specific default value, typically the last-modified-time
or a FileTime representing the epoch (1970-01-01T00:00:00Z).
```

So there is no change there. Even on some Linux systems where `statx` isn't supported you'd get the old behaviour. This is according to spec.

File: test/jdk/java/nio/file/attribute/BasicFileAttributeView/CreationTime.java [hunk 1]
```diff
@@ -81,7 +83,13 @@ static void test(Path top) throws IOException {
                 supportsCreationTimeRead = true;
                 supportsCreationTimeWrite = true;
             }
+        } else if (Platform.isLinux()) {
+            // Creation time read depends on statx system call support
+            supportsCreationTimeRead = CreationTimeHelper.linuxIsCreationTimeSupported();
+            // Creation time updates are not supported on Linux
+            supportsCreationTimeWrite = false;
         }
+        System.out.println("supportsCreationTimeRead == " + supportsCreationTimeRead);
```
[the remark this answers is not in the document]
Comment (jerboaa) [reply test/jdk/java/nio/file/attribute/BasicFileAttributeView/CreationTime.java:92]: I'm not sure that's really worth it. Sleep isn't elegant either. Anyway, nothing for a backport to do.

**Comment by jerboaa** [comment]
macos-x64 GHA failure of `gc/logging/TestMetaSpaceLog` seems unrelated.

**Comment by tstuefe** [comment]
> Thanks for the review!
> 
> > I wonder about backward compatibility - whether there is code that relies on creation time being the modification time. If so, its an error of course, but still may cause problems.
> 
> Users already get different behaviour on say, Mac OS, Windows and Linux (with statx support) today. The former two support `creationTime()`, Linux does not (prior this patch). The [javadoc](https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/nio/file/attribute/BasicFileAttributes.html#creationTime()) from creationTime() mentions:
> 
> ```
> If the file system implementation does not support a time stamp to indicate the time when the file was
> created then this method returns an implementation specific default value, typically the last-modified-time
> or a FileTime representing the epoch (1970-01-01T00:00:00Z).
> ```
> 
> So there is no change there. Even on some Linux systems where `statx` isn't supported you'd get the old behaviour. This is according to spec.

Okay. In the end, you are the maintainer, so if the risk is acceptable to you it is to me too.

**Comment by jerboaa** [comment]
> > Users already get different behaviour on say, Mac OS, Windows and Linux (with statx support) today. The former two support creationTime(), Linux does not (prior this patch).
> 
> It looks like the behavior of `BasicFileAttributes` and `BasicFileAttributeView` is now inconsistent on Linux (Apline). I can change the time using:
> 
> ```java
> BasicFileAttributeView attributes = Files.getFileAttributeView(path, BasicFileAttributeView.class);
> attributes.setTimes(time, time, time);
> ```
> 
> And the 3d param is called `createTime`. But apparently it sets a different attribute than what we get using:
> 
> ```java
> FileTime actual = Files.readAttributes(path, BasicFileAttributes.class).creationTime();
> ```
> 
> Tried with OpenJDK 24.

Updating of the creation time is not supported on Linux as far as I know. So you'd get the creation time when the file was originally created, not when `setTimes()` has been called.

**Comment by ctapobep** [comment]
My point is that the methods used to be consistent: you pass createTime when changing it through `BasicFileAttributeView`, and you get the same result in `BasicFileAttributes.creationTime()`. I don't know how this worked before, but it did. I think it makes sense to keep the API consistent, or at least mention the inconsistencies in the docs.

**Comment by ctapobep** [comment]
Yes, that's what I ended up doing. But it took me quite some time to research this topic. I'm just asking to make the API easier to understand for the next person who runs into this.

It's a particularly nasty problem because it was working fine on my MacOS, and it failed only on Linux.

**Comment by GoeLin** [comment]
> > My point is that the methods used to be consistent: you pass createTime when changing it through `BasicFileAttributeView`, and you get the same result in `BasicFileAttributes.creationTime()`. I don't know how this worked before, but it did. I think it makes sense to keep the API consistent, or at least mention the inconsistencies in the docs.
> 
> There was a release note: https://bugs.openjdk.org/browse/JDK-8331874
> 
> It "worked" before by accident. Creation time returned the last modified time before. Use last modified time (over creation time) and you should have the same consistency as before.

The release note does not mention 17, though.

**Pull request closed without merging**